File: algorithms/java/最大子数组和/最大子数组和.java
// Source : https://leetcode.cn/problems/maximum-subarray/
// Author : Yanan Wang
// Date   : 2023-06-23

class Solution {
    public int maxSubArray(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n + 1];
        int result = nums[0];
        for (int i = 1;  i <= n; i++) {
            dp[i] = Math.max(nums[i - 1], dp[i - 1] + nums[i - 1]);
            result = Math.max(result, dp[i]);
        }
        return result;
    }
}

File: algorithms/python/最大子数组和/最大子数组和.py
# Source : https://leetcode.cn/problems/maximum-subarray/description/
# Author : Yanan Wang
# Date   : 2024-01-16

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0] * n
        dp[0] = nums[0]

        for i in range(1, n):
            dp[i] = max(dp[i - 1] + nums[i], nums[i])
            
        ans = dp[0]
        for i in range(1, n):
            ans = max(ans, dp[i])
        
        return ans